src/server/coprocessor.rs

```diff
@@ -286,132 +277,195 @@ fn is_point(range: &KeyRange) -> bool {
+                } else {
+                    let raw_key = table::encode_column_key(tid, h, col.get_column_id());
+                    let key = Key::from_raw(raw_key);
+                    match try!(self.snap.get(&key)) {
+                        None => return Err(box_err!("key {:?} not exists", key)),
+                        Some(bs) => row.mut_data().extend(bs),
+                    }
+                }
+            }
         }
+        row.set_handle(handle);
+        Ok(Some(row))
     }
-    let res = box_try!(eval.eval(sel.get_field_where()));
-    if let Datum::Null = res {
-        trace!("got null, skip.");
-        return Ok(None);
-    }
-    if box_try!(res.as_bool()) {
-        trace!("pass.");
-        return Ok(Some(row));
-    }
-    trace!("got false, skip.");
-    Ok(None)
-}
 
-fn get_rows_from_idx(snap: &SnapshotStore,
-                     sel: &SelectRequest,
-                     ranges: Vec<KeyRange>)
-                     -> Result<Vec<Row>> {
-    let mut rows = vec![];
-    for r in ranges {
-        let part = try!(get_idx_row_from_range(snap, sel.get_index_info(), r));
-        rows.extend(part);
+    fn get_rows_from_idx(&self, ranges: Vec<KeyRange>) -> Result<Vec<Row>> {
+        let mut rows = vec![];
+        for r in ranges {
+            let part = try!(self.get_idx_row_from_range(r));
+            rows.extend(part);
+        }
+        Ok(rows)
     }
-    Ok(rows)
-}
 
-fn get_idx_row_from_range(snap: &SnapshotStore,
-                          info: &IndexInfo,
-                          mut r: KeyRange)
-                          -> Result<Vec<Row>> {
-    let mut rows = vec![];
-    let mut seek_key = r.take_start();
-    loop {
-        trace!("seek {:?}", seek_key);
-        let mut nk = try!(snap.scan(Key::from_raw(seek_key.clone()), 1));
-        if nk.is_empty() {
-            debug!("no more data to scan");
-            return Ok(rows);
-        }
-        let (key, value) = try!(nk.pop().unwrap());
-        if r.get_end() <= &key {
-            debug!("reach end key: {:?} >= {:?}", key, r.get_end());
-            return Ok(rows);
+    fn get_idx_row_from_range(&self, mut r: KeyRange) -> Result<Vec<Row>> {
+        let mut rows = vec![];
+        let info = self.sel.get_index_info();
+        let mut seek_key = r.take_start();
+        loop {
+            trace!("seek {:?}", seek_key);
+            let mut nk = try!(self.snap.scan(Key::from_raw(seek_key.clone()), 1));
+            if nk.is_empty() {
+                debug!("no more data to scan");
+                return Ok(rows);
+            }
+            let (key, value) = try!(nk.pop().unwrap());
+            if r.get_end() <= &key {
+                debug!("reach end key: {:?} >= {:?}", key, r.get_end());
+                return Ok(rows);
+            }
+            let mut datums = box_try!(table::decode_index_key(&key));
+            let handle = if datums.len() > info.get_columns().len() {
+                datums.pop().unwrap()
```

Comment:
Because handle must be in the value or appended to the key. Here we detect which case is encountered.